### PR TITLE
Enabling Membership resource to be regionalized with global as default value

### PR DESCRIPTION
**Pull request by sahsagar-google**
https://github.com/hashicorp/terraform-provider-google/issues/15882

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
gkehub: added `location` field to `google_gke_hub_membership` resource
```

```release-note:enhancement
gkehub: added `location` field to `google_gke_hub_membership_iam_*` resources
```

### CODE REVIEW

**Review by melinath — CHANGES_REQUESTED** [review]
It looks like this might be a breaking change for the IAM resources; if so, fixing it will require adding `schema_version` / `state_upgraders`, which are not currently supported by iam resources. I'll need to investigate some more.

File: mmv1/products/gkehub/Membership.yaml [hunk 1]
```diff
@@ -60,6 +60,16 @@ examples:
     vars:
       name: 'basic'
       cluster_name: 'basiccluster'
+  - !ruby/object:Provider::Terraform::Examples
+    name: 'gkehub_membership_regional'
```
[the remark this answers is not in the document]
Comment (sahsagar-google) [reply mmv1/products/gkehub/Membership.yaml]: Ah, I see. Please let me know if the new changes work. Thank you.

**Comment by melinath** [comment]
confirmed this is a breaking change for the IAM resources. Error:

```
Error when reading or editing GKEHubMembership "projects/project-id/locations/global/memberships/basic": googleapi: Error 400: Malformed name: 'projects/project-id/locations//memberships/basic
```

**Comment by melinath** [comment]
I'll work on a workaround.

**Comment by melinath** [comment]
I have a tentative implementation at https://github.com/GoogleCloudPlatform/magic-modules/pull/9035 but it's not quite working yet. I'll be able to look into it more next week. Apologies for the delay.

File: mmv1/products/gkehub/Membership.yaml [hunk 2]
```diff
@@ -53,6 +53,16 @@ iam_policy: !ruby/object:Api::Resource::IamPolicy
     ]
 autogen_async: true
```
Comment (melinath) [inline mmv1/products/gkehub/Membership.yaml:54]:
```suggestion
autogen_async: true
schema_version: 1
state_upgraders: true
```

**Review by melinath — CHANGES_REQUESTED** [review]
Update: it looks like a state upgrader is not actually required for the IAM resources. However, you will need to add a state upgrader for the GKEHubMembership resource - that means adding a couple lines to Membership.yaml and a state migration file like https://github.com/GoogleCloudPlatform/magic-modules/pull/9035/files#diff-a9abd9e5a363f02593a71489617efd5fc180e880bf3c360c22eca093c5836585

You can test that this works by doing [manual tests](https://googlecloudplatform.github.io/magic-modules/develop/run-tests/#run-manual-tests) to create a membership with this PR's main branch base commit, then generating the provider using this PR, building it, and trying to run a `terraform plan`. Sample config:

```
resource "google_container_cluster" "primary" {
  name               = "basiccluster"
  location           = "us-central1-a"
  initial_node_count = 1
}

resource "google_gke_hub_membership" "membership" {
  membership_id = "basic"
  # location = "%{location}"
  endpoint {
    gke_cluster {
      resource_link = "//container.googleapis.com/${google_container_cluster.primary.id}"
    }
  }
}
```

**Review by melinath — CHANGES_REQUESTED** [review]
Failure is:

```
No such file or directory @ rb_sysopen - /mm-terraform-ga-head/mmv1/templates/terraform/state_migrations/gke_hub_membership.go.erb
```

This is due to needing a state upgrader - I think the file at https://github.com/GoogleCloudPlatform/magic-modules/pull/9035/files#diff-a9abd9e5a363f02593a71489617efd5fc180e880bf3c360c22eca093c5836585 should work for you.

File: mmv1/templates/terraform/examples/gkehub_membership_regional.tf.erb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+resource "google_container_cluster" "primary" {
+  name               = "<%= ctx[:vars]['cluster_name'] %>"
+  location           = "us-central1-a"
+  initial_node_count = 1
```
Comment (melinath) [inline mmv1/templates/terraform/examples/gkehub_membership_regional.tf.erb:4]: In 5.0.0, google_container_cluster.deletion_protection started [defaulting to `true`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/version_5_upgrade#cluster-deletion-now-prevented-by-default-with-deletion_protection) so we need to explicitly set it to `false` here.

If you don't want to include this in the docs, you could use a var to set the value.

```suggestion
  initial_node_count = 1
  deletion_protection = false
```

**Review by melinath — CHANGES_REQUESTED** [review]
LGTM overall - state migrations seem to work as expected & resolve the upgrade issue. I think the tgc integration test failures could be resolved by merging the main branch into this PR (or rebasing.) Most of the VCR failures seem unrelated - added a note below about resolving the remainder.

**Review by melinath — APPROVED** [review]
LGTM

**Comment by melinath** [comment]
TGC failures seem unrelated - I believe they have been resolved on main.

**Pull request merged**